doc/developer/design/20231213_compute_operator_hydration_status_tracking.md

```diff
@@ -0,0 +1,192 @@
+* **Technical Feasibility:** Attaching hydration monitoring to every dataflow operator is difficult given the current code structure.
+  A good solution would likely require us to implement support into the Timely Dataflow runtime.
+
+Instead, this design proposes to only report hydration status of dataflow regions that correspond to LIR nodes.
+LIR is the low-level intermediate representation used by Materialize's optimizer.
+Users can show their queries as LIR plans using the `EXPLAIN PHYSICAL PLAN` syntax.
+LIR plans are the lowest level of query representation we document and that (advanced) users can therefore be expected to understand.
+LIR nodes map directly to groups of operators in a dataflow, so it is possible for us to collect hydration status information at this level.
+
+While it would be desirable to also provide hydration progress information at a higher level, we consider that outside the scope of this design.
+
+Reporting LIR-level hydration status requires a number of pieces that don't exist yet in our code:
+
+1. [Assigning IDs to LIR nodes during query planning](#lir-node-ids)
+2. [Creating a separate region for each LIR node during Compute rendering](#lir-dataflow-regions)
+3. [Collecting LIR region hydration status during dataflow processing](#hydration-status-logging)
+4. [Reporting LIR region hydration status to the controller](#reporting-introspection-data)
+
+While there is some amount of lifting required to implement all these pieces, our hope is that they will prove generally useful.
+(1) and (2) together allow mapping low-level dataflow information back to `EXPLAIN` plans and will be useful for making our internal memory visualizer user-facing.
+They are also a precondition for surfacing runtime information, like memory usage, in `EXPLAIN` plans.
+(4) gives us a reliable way to expose dataflow introspection information in SQL, which will be useful in alleviating some of the pain currently caused by the bad ergonomics of our per-replica introspection mechanism.
+
+### LIR Node IDs
+
+To make LIR nodes identifiable, we assign IDs to them (referred to as _LIDs_ in the following).
+LIDs need to be unique only within a dataflow, as they can be namespaced by the dataflow ID, or by the `GlobalId` of the SQL object that created the dataflow.
+
+This design proposes that LIDs should be increasing integers.
+This is the simplest ID format that introduces the least visual noise when exposed in, e.g., `EXPLAIN` plans, and there is currently no need for anything more complex.
+Since both the `EXPLAIN` output and the introspection tables that will expose operator hydration information are part of Materialize's unstable API, we are free to change the ID format later, should the need arise.
+
+LIDs need to be assigned as part of query planning.
+Because LIR-level optimizations can change the plan by removing and adding nodes, ID assignment should ideally happen after all optimization have been performed.
```

Comment:
> Because LIR-level optimizations can change the plan by removing and adding nodes, ID assignment should ideally happen after all optimization have been performed.

This is true, but:

- AFAIK currently we don't have optimizations that add or remove LIR nodes (they only refine LIR node properties).
- If we want to expose hydration (or any other derived) information on top of the (MIR-based) `OPTIMIZED PLAN` as opposed to the (LIR-based) `PHYSICAL PLAN`, we will need to create an association map between MIR nodes and LIR nodes.

In the long term it is therefore probably better if we assign the nodes when creating the LIR nodes as part of the `Self::lower_dataflow` call in `finalize_dataflow`.

Reply:
> AFAIK currently we don't have optimizations that add or remove LIR nodes (they only refine LIR node properties).

I wasn't entirely sure about this, but you are probably right.

Allocating IDs when the LIR nodes are created would simplify the types somewhat, I think (`Plan`s would always have `NodeId` types, not sometimes `()` instead). Let me give implementing this variant a shot.

Reply:
After thinking some more about this, I don't think your proposal is a blocker.

The crucial think is to be able to keep track of the MIR-level ID under which the LIR node was generated in-line (as part of the LIR variant) when we do MIR ⇒ LIR lowering. When and how we complete the association by assigning unique LIR nodes is not so important I guess.

Reply:
I amended the design to propose assigning the IDs during lowering, since that is easy to implement and keeps the types simpler.